crates/iota-network/src/discovery/server.rs

```diff
@@ -56,54 +54,30 @@ impl Discovery for Server {
             .clone()
             .ok_or_else(|| anemo::rpc::Status::internal("own_info has not been initialized yet"))?;
 
-        let known_peers = if state.known_peers.len() < MAX_PEERS_TO_SEND {
-            state
-                .known_peers
-                .values()
-                .map(|e| e.inner())
-                .cloned()
-                .collect()
-        } else {
-            let mut rng = rand::thread_rng();
-            // prefer returning peers that we are connected to as they are known-good
-            let mut known_peers = state
-                .connected_peers
-                .keys()
-                .filter_map(|peer_id| state.known_peers.get(peer_id))
-                .map(|info| (info.peer_id, info))
-                .choose_multiple(&mut rng, MAX_PEERS_TO_SEND)
-                .into_iter()
-                .collect::<HashMap<_, _>>();
-
-            if known_peers.len() <= MAX_PEERS_TO_SEND {
-                // Fill the remaining space with other peers, randomly sampling at most
-                // MAX_PEERS_TO_SEND
-                for info in state
-                    .known_peers
-                    .values()
-                    // This randomly samples the iterator stream but the order of elements after
-                    // sampling may not be random, this is ok though since we're just trying to do
-                    // best-effort on sharing info of peers we haven't connected with ourselves.
-                    .choose_multiple(&mut rng, MAX_PEERS_TO_SEND)
-                {
-                    if known_peers.len() >= MAX_PEERS_TO_SEND {
-                        break;
-                    }
-
-                    known_peers.insert(info.peer_id, info);
-                }
-            }
-
-            known_peers
-                .into_values()
-                .map(|e| e.inner())
-                .cloned()
-                .collect()
-        };
+        let mut rng = rand::thread_rng();
+        let mut known_connected_peers = state
+            .known_peers
+            .iter()
+            .filter_map(|(peer_id, peer_info)| {
+                (peer_info.access_type != AccessType::Private
+                    && state.connected_peers.contains_key(peer_id))
+                .then_some(peer_info.inner().clone())
+            })
+            .choose_multiple(&mut rng, MAX_PEERS_TO_SEND);
+        let mut known_not_connected_peers = state
+            .known_peers
+            .iter()
+            .filter_map(|(peer_id, peer_info)| {
+                (peer_info.access_type != AccessType::Private
+                    && !state.connected_peers.contains_key(peer_id))
+                .then_some(peer_info.inner().clone())
+            })
+            .choose_multiple(&mut rng, MAX_PEERS_TO_SEND - known_connected_peers.len());
+        known_connected_peers.append(&mut known_not_connected_peers);
```

Comment:
Should we maybe keep the old name "known_peers" instead? The variable contains connected and unconnected known peers maybe.

Reply:
Should we shuffle the final list, so that our peers can't directly determine to which peers we are connected?

Reply:
Right, that wouldn't hurt.